test/e2e/kubectl.go

```diff
@@ -0,0 +1,201 @@
+
+	It("should create and stop a replication controller", func() {
+		defer cleanup(nautilusPath)
+
+		By("creating a replication controller")
+		runKubectl("create", "-f", nautilusPath)
+		validateController(nautilusImage, 2, 30*time.Second)
+	})
+
+	It("should scale a replication controller", func() {
+		defer cleanup(nautilusPath)
+
+		By("creating a replication controller")
+		runKubectl("create", "-f", nautilusPath)
+		validateController(nautilusImage, 2, 30*time.Second)
+		By("scaling down the replication controller")
+		runKubectl("resize", "rc", "update-demo-nautilus", "--replicas=1")
+		validateController(nautilusImage, 1, 30*time.Second)
+		By("scaling up the replication controller")
+		runKubectl("resize", "rc", "update-demo-nautilus", "--replicas=2")
+		validateController(nautilusImage, 2, 30*time.Second)
+	})
+
+	It("should do a rolling update of a replication controller", func() {
+		// Cleanup all resources incase we fail somewhere in the middle
+		defer cleanup(updateDemoRoot)
+
+		By("creating the initial replication controller")
+		runKubectl("create", "-f", nautilusPath)
+		validateController(nautilusImage, 2, 30*time.Second)
+		By("rollingupdate to new replication controller")
+		runKubectl("rollingupdate", "update-demo-nautilus", "--update-period=1s", "-f", kittenPath)
+		validateController(kittenImage, 2, 30*time.Second)
+	})
+
+})
+
+func cleanup(filePath string) {
+	By("using stop to clean up resources")
+	runKubectl("stop", "-f", filePath)
+
+	resources := runKubectl("get", "pods,rc", "-l", updateDemoSelector, "--no-headers")
+	if resources != "" {
+		Failf("Resources left running after stop:\n%s", resources)
+	}
+}
+
+func validateController(image string, replicas int, timeout time.Duration) {
+
+	getPodsTemplate := "--template={{range.items}}{{.id}} {{end}}"
+
+	// NB: kubectl adds the "exists" function to the standard template functions.
+	// This lets us check to see if the "running" entry exists for each of the containers
+	// we care about. Exists will never return an error and it's safe to check a chain of
+	// things, any one of which may not exist. In the below template, all of info,
+	// containername, and running might be nil, so the normal index function isn't very
+	// helpful.
+	// This template is unit-tested in kubec{tl|fg}, so if you change it, update the unit test.
+	//
+	// You can read about the syntax here: http://golang.org/pkg/text/template/
+	getContainerStateTemplate := fmt.Sprintf(`--template={{and (exists . "currentState" "info" "%s" "state" "running")}}`, updateDemoContainer)
+
+	getImageTemplate := fmt.Sprintf(`--template={{(index .currentState.info "%s").image}}`, updateDemoContainer)
+
+	getHostIPTemplate := "--template={{.currentState.hostIP}}"
+
+	By(fmt.Sprintf("waiting for all containers in %s pods to come up.", updateDemoSelector))
+	for start := time.Now(); time.Since(start) < timeout; time.Sleep(5 * time.Second) {
+		getPodsOutput := runKubectl(
```

Comment:
Oh dear, my python is showing :P. Changed to single lines, but kept them as distinct args. Note that `exec.Command` takes `..string`, so if `runKubectl`takes a single string it would need to split it (The join in there is purely for logging purposes), which seems silly.